<a href="https://colab.research.google.com/github/LuFernandaA/Boomcamp-Data-Sciense/blob/main/TP1/Actividad_2_Limpieza_y_An%C3%A1lisis_de_Datos_con_Pandas_y_NumPy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad 2: Limpieza y Análisis de Datos con Pandas y NumPy**

In [ ]:
#Importamos pandas
import pandas as pd

#Importamos numpy
import numpy as np

In [ ]:
#Importamos la base de datos
from google.colab import drive, files
uploaded = files.upload()

Saving ventas_sucias_5000.csv to ventas_sucias_5000 (1).csv


**Parte 1 — Exploración de datos (Pandas)**📌

In [ ]:
#Cargo la base de datos
df = pd.read_csv('/content/ventas_sucias_5000 (1).csv')

In [ ]:
#Miramos primeras filas
df.head()

,cliente,producto,precio,cantidad,pais,metodo_pago,fecha
0,Maria,Monitor,1326.0,NaN,peru,Efectivo,2024-01-01 00:00:00
1,Luisa,Laptop,55.0,2,chile,Tarjeta,2024-01-01 01:00:00
2,Carlos,Monitor,1203.0,9,Colombia,Efectivo,2024-01-01 02:00:00
3,Luisa,Monitor,1304.0,3,Perú,TRANSFERENCIA,2024-01-01 03:00:00
4,Luisa,Monitor,426.0,6,chile,Tarjeta,2024-01-01 04:00:00


In [ ]:
#Vista previa de la información general del dataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   cliente      5000 non-null   object 
 1   producto     5000 non-null   object 
 2   precio       4950 non-null   float64
 3   cantidad     4950 non-null   object 
 4   pais         5000 non-null   object 
 5   metodo_pago  5000 non-null   object 
 6   fecha        5000 non-null   object 
dtypes: float64(1), object(6)
memory usage: 273.6+ KB


In [ ]:
#Observamos el resumen estadístico
df.describe()

,precio
count,4950.000000
mean,5053.823434
std,63379.982009
min,10.000000
25%,531.000000
50%,1027.000000
75%,1519.000000
max,999999.000000


In [ ]:
#Miro cuantas filas y columnas tengo
df.shape

(5000, 7)

In [ ]:
#Identifico valores inconsistentes en pais
df['pais'].unique()

array(['peru', 'chile', 'Colombia', 'Perú', 'COL', 'Chile', 'col'],
      dtype=object)

In [ ]:
#Identifico valores inconsistentes en metodo_pago
df['metodo_pago'].unique()

array(['Efectivo', 'Tarjeta', 'TRANSFERENCIA', 'transferencia'],
      dtype=object)

In [ ]:
#Identifico valores inconsistentes en cantidad
df['cantidad'].unique()

array([nan, '2', '9', '3', '6', '7', '1', '8', '5', '4', 'three'],
      dtype=object)

In [ ]:
#Identifico faltantes
df.isnull().sum()

,0
cliente,0
producto,0
precio,50
cantidad,50
pais,0
metodo_pago,0
fecha,0


Precio y cantidad le hacen falta 50 datos.

In [ ]:
#Identifico duplicados
df.duplicated().sum()

np.int64(0)

No se encuentran datos repetidos.

**¿Cuántas filas y columnas tiene la base de datos?**

Gracias a la vista previa y el codigo df.shape, podemos determinar que en total tenemos 5000 filas en 7 columnas.


**¿Qué tipos de datos identificas?**

Cadena de caracteres (object), Formato de decimal (float64).


**¿Encuentras columnas con problemas o inconsistencias?**

Si, las columnas donde pude identificar inconsistencias es la cantidad y fecha, por ende, debo cambiar el tipo de dato.

La columna de pais peru aparece con tilde, sin tilde y en colombia, con C mayuscula, minuscula y abreviada como COL o col. Y chile con C en mayuscula y minuscula. Adicional, en metodo de pago tambien pude identificar algunos en mayusculas y otros en minusculas.

Y por ultimo, no todas las columnas tienen la misma cantidad de datos.

In [ ]:
#Creo una copia para proteger mi base de datos original
df1 = df.copy()

**Parte 2 — Limpieza de datos (Pandas)**📌

In [ ]:
df1

,cliente,producto,precio,cantidad,pais,metodo_pago,fecha
0,Maria,Monitor,1326.0,NaN,peru,Efectivo,2024-01-01 00:00:00
1,Luisa,Laptop,55.0,2,chile,Tarjeta,2024-01-01 01:00:00
2,Carlos,Monitor,1203.0,9,Colombia,Efectivo,2024-01-01 02:00:00
3,Luisa,Monitor,1304.0,3,Perú,TRANSFERENCIA,2024-01-01 03:00:00
4,Luisa,Monitor,426.0,6,chile,Tarjeta,2024-01-01 04:00:00
...,...,...,...,...,...,...,...
4995,Juan,Laptop,1582.0,8,chile,transferencia,2024-07-27 03:00:00
4996,Carlos,Mouse,1553.0,2,Perú,Tarjeta,2024-07-27 04:00:00
4997,Juan,Monitor,1776.0,2,chile,Efectivo,2024-07-27 05:00:00
4998,Pedro,Laptop,1366.0,8,Chile,TRANSFERENCIA,2024-07-27 06:00:00


In [ ]:
#Convierto vacios y Estandarizo/Mapeo la columna de pais

df1['pais'] = df1['pais'].str.lower()

df1['pais'] = df1['pais'].replace("",None)
df1['pais'] = df1['pais'].replace(" ",None)

df1['pais'] = df1['pais'].replace(
    {"col":"Colombia",
     "COL":"Colombia",
     "colombia":"Colombia",
     "peru":"Peru",
     "perú":"Peru",
     "chile":"Chile"}
)

#Estandarizo/Mapeo la columna de metodo_pago

df1['metodo_pago'] = df1['metodo_pago'].str.lower()
df1['metodo_pago'] = df1['metodo_pago'].replace({
    "transferencia":"Transferencia",
    "efectivo":"Efectivo",
    "tarjeta":"Tarjeta",
    "TRANFERENCIA":"Transferencia"})


df1['metodo_pago'] = df1['metodo_pago'].replace("",None)
df1['metodo_pago'] = df1['metodo_pago'].replace(" ",None)

In [ ]:
#Manejo nulos en precio

df1['precio'] = df1['precio'].replace("",None)
df1['precio'] = df1['precio'].replace(" ",None)
df1['precio'] = df1['precio'].fillna(df1['precio'].mean())

In [ ]:
#Convierto, corrijo y manejo nulos numericos en la columna de cantidad

df1['cantidad'] = pd.to_numeric(df1['cantidad'], errors='coerce')
df1['cantidad'] = df1['cantidad'].fillna(df1['cantidad'].mean())
df1['cantidad'] = df1['cantidad'].round(1)

In [ ]:
#Convierto columna fecha a formato de fecha

df1['fecha'] = pd.to_datetime(df1['fecha'],errors='coerce')
df1 = df1.dropna(subset=['fecha'])

In [ ]:
df1.head(50) #verifico si mi limpieza se realizo de la manera correcta

,cliente,producto,precio,cantidad,pais,metodo_pago,fecha
0,Maria,Monitor,1326.0,5.0,Peru,Efectivo,2024-01-01 00:00:00
1,Luisa,Laptop,55.0,2.0,Chile,Tarjeta,2024-01-01 01:00:00
2,Carlos,Monitor,1203.0,9.0,Colombia,Efectivo,2024-01-01 02:00:00
3,Luisa,Monitor,1304.0,3.0,Peru,Transferencia,2024-01-01 03:00:00
4,Luisa,Monitor,426.0,6.0,Chile,Tarjeta,2024-01-01 04:00:00
5,Juan,Laptop,857.0,6.0,Colombia,Transferencia,2024-01-01 05:00:00
6,Carlos,Laptop,148.0,2.0,Chile,Tarjeta,2024-01-01 06:00:00
7,Carlos,Celular,256.0,7.0,Chile,Tarjeta,2024-01-01 07:00:00
8,Carlos,Mouse,1181.0,9.0,Chile,Tarjeta,2024-01-01 08:00:00
9,Luisa,Mouse,239.0,1.0,Chile,Efectivo,2024-01-01 09:00:00


In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   cliente      5000 non-null   object        
 1   producto     5000 non-null   object        
 2   precio       5000 non-null   float64       
 3   cantidad     5000 non-null   float64       
 4   pais         5000 non-null   object        
 5   metodo_pago  5000 non-null   object        
 6   fecha        5000 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(2), object(4)
memory usage: 273.6+ KB


In [ ]:
df1.isnull().sum()

,0
cliente,0
producto,0
precio,0
cantidad,0
pais,0
metodo_pago,0
fecha,0


**¿Qué problemas encontraste?**

Encontre problemas en el metodo de pago, cantidad, fecha y algunos datos faltantes.

**¿Cómo los solucionaste?**
Comence a estandarizar, convertir y manejar todos esos datos nulos con los codigos que nos indico la profesora viviana en la clase.

**¿Eliminaste registros? ¿Por qué?**
Como tal no tuve que eliminar registros, pero si tuve que hacer un par de cambios para que mi base de datos quedara limpia. Utilice la media para cantidad, y asi no dejar esos datos vacios.

**Parte 3 — Análisis con Pandas**📌




In [ ]:
#Añado nueva columna llamada total: df['total'] = df['precio'] * df['cantidad']

df1["total"] = df1["precio"] * df1["cantidad"]

In [ ]:
#Total de ventas
Total_ventas = df1['total'].sum()
print(Total_ventas)

103289535.67878786


In [ ]:
#Promedio de ventas
Promedio_ventas = df1['total'].mean()
print(Promedio_ventas)

20657.90713575757


In [ ]:
#Venta maxima
Venta_maxima = df1['total'].max()
print(Venta_maxima)

8999991.0


In [ ]:
#Venta minima
Venta_minima = df1['total'].min()
print(Venta_minima)

10.0


In [ ]:
#Productos con mayor ventas
Productos_mayor_ventas = df1.groupby('producto')['total'].sum().sort_values(ascending=False)
pd.set_option('display.float_format', lambda x: '%.2f' % x)
print(Productos_mayor_ventas)

producto
Mouse     27672331.17
Teclado   27016888.70
Laptop    18330792.35
Monitor   16209749.88
Celular   14059773.58
Name: total, dtype: float64


In [ ]:
#Paises con más ventas
Pais_ventas = df1.groupby('pais')['total'].sum().sort_values(ascending=False)
pd.set_option('display.float_format', lambda x: '%.2f' % x)
print(Pais_ventas)

pais
Colombia   49398894.22
Chile      28584713.46
Peru       25305927.99
Name: total, dtype: float64


**Parte 4 — Introducción a NumPy**📌

In [ ]:
#Convertir columnas numéricas a un array de NumPy
Total_ventas = df1[['precio', 'cantidad']].to_numpy()

In [ ]:
#Separar columnas utilizando indexación
precios = data[:, 0]
cantidades = data[:, 1]

In [ ]:
#Vectorización para calcular ventas totales
Total_ventas = precios * cantidades
print(Total_ventas)

print("Total_ventas:",np.sum(Total_ventas))

[ 6630.   110. 10827. ...  3552. 10928.   406.]
Total_ventas: 103289535.67878786


**Parte 5 — Análisis con NumPy**📌

In [ ]:
#Suma total de ventas
print("Total_ventas:",np.sum(Total_ventas))

Total_ventas: 103289535.67878786


In [ ]:
#Promedio de ventas
print("Promedio de ventas:", np.mean(Total_ventas))

Promedio de ventas: 20657.90713575757


In [ ]:
#Venta maxima
print("Venta_maxima:", np.max(Total_ventas))

Venta_maxima: 8999991.0


In [ ]:
#Cantidad de ventas superiores a 1000
print("Cantidad de ventas superiores a 1000:", np.sum(Total_ventas > 1000))

Cantidad de ventas superiores a 1000: 4256


**¿Qué ventajas observas al usar NumPy?**

Al usar numpy pude observar que todo se agiliza y es más sencillo, no es necesario darle las variables, solo comenzar a indicarle si quiero la maxima, minima, la media, entre otras operaciones de una manera más rapida y sin tantas complicaciones.

**¿Qué significa “vectorización”?**

La vectorización en numpy es aplicar operaciones sobre un array, el cual operara sobre todos los elementos de manera rapida y eficiente.

**¿Qué hace la expresión data[:, 0]?**

Esta se aplica para indexación y permite elegir todas las filas desde 0.

**Parte 6 — Interpretación de resultados**📌

**¿Los resultados obtenidos tienen sentido?**
Si, ya que me permitieron identificar cuales eran la media, el total de ventas, entre otros datos necesarios para llegar a un analisis.

**¿Detectaste valores sospechosos?**

Al momento de hacer la exploración y la limpieza de datos no identifique como tal algun valor sospechoso. Si habia unos NaN, pero trate con ellos en la limpieza.

**¿El promedio representa correctamente los datos?**

Si, el promedio que me arroja numpy es correcto **20657.90713575757** ya que este debe ser la división del total y las filas 103,289,535.67 / 5000

**¿Qué decisiones tomarías si esta fuera información real de negocio?**

Lo primero seria enfocarme los productos más vendidos y menos vendidos,definiria estrategias para negociar mejores precios con los proveedores y revisar si los precios son competitivos. Crear una manera de llegar a más personas en otros paises con promociones y impulsando las ventas, sin dejar de lado colombia ya que domina el 50% de las ventas.